### PR TITLE
[editorial] Remove unreachable branch

### PR DESCRIPTION
**Pull request by jugglinmike**

### CODE REVIEW

**Comment by bterlson** [comment]
Is it just me or are there many more instances of this across the spec?

**Comment by jugglinmike** [comment]
I'm sure it's possible, but this is the first I've identified.

**Comment by jugglinmike** [comment]
Although I suppose gh-476 is essentially the same class of change. It's a little more subtle because the unreachable branch is represented by a single character.

**Comment by bterlson** [comment]
@jugglinmike IIUC, this should also apply to many of the other callsites of IteratorDestructuringAssignmentEvaluation. Eg. 12.15.5.2 contains two calls of the exact same form. Thoughts?

**Comment by jugglinmike** [comment]
I see what you mean. Looking at this more closely, I do believe these are the
only two instances of this particular oversight. The change generally concerns
any ArrayAssignmentPattern which contains an Elision followed by any other
production, and that only occurs in these two parts of the specification.

I've added a second commit to address the other case.

Destructuring *binding* seemed a likely location for additional changes along
these lines... Except that closing semantics are handled much differently there
(and more neatly, I think). Instead of creating an "iteratorRecord" in every
specific ArrayBindingPattern production (and re-implementing closing
semantics), it does this generically for *all* ArrayBindingPatterns:

>     BindingPattern : ArrayBindingPattern
>
>     1. Let iterator be ? GetIterator(value).
>     2. Let iteratorRecord be Record {[[Iterator]]: iterator, [[Done]]:
>        false}.
>     3. Let result be IteratorBindingInitialization for ArrayBindingPattern
>        using iteratorRecord and environment as arguments.
>     4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator,
>        result).
>     5. Return result.

Expressed directly in code, this might be slightly less efficient because:

1. You don't need a record for the "empty" pattern (`[]`)
2. As we've been discussing, there are some specific cases where it is not
   necessary to check [[Done]]

But the intent is much more clear, and maybe that's what's important here; we
can let the engines worry about optimizations.

This  makes me wonder: what would you think about reformatting the iterator
closing language for ArrayAssignmentPattern to match ArrayBindingPattern? I'm
happy to submit a patch to demonstrate, but it would help me if I knew that
there is interest in such a change.

**Comment by bterlson** [comment]
I am interested in such a change (as evidenced above I find this hard to follow). In the meantime, I will take this PR!
